algosdk/template.py

```diff
@@ -0,0 +1,434 @@
+
+        contract : bytes
+            the contract contaning information, should be recieved from payer
+        seceret_key : bytes
+            the secret key to sign the contract
+        """
+        pass
+
+
+class LimitOrder(Template):
+    """
+    Limit Order allows to trade Algos for other asests given a specific ratio;
+    for N Algos, swap for Rate * N Assets.
+    ...
+
+    Args:
+        owner (str): an address that can receive the asset after the expiry
+            round
+        ratn (int): the numerator of the exchange rate
+        ratd (int): the denominator of the exchange rate
+        expiry_round (int): the round on which the assets can be transferred
+            back to owner
+        min_trade (int): the minimum amount (of Algos) to be traded away
+        max_fee (int): the maximum fee that can be paid to the network by the
+            account
+
+    """
+    def __init__(self, owner: str, ratn: int, ratd: int, expiry_round: int,
+                 min_trade: int, max_fee: int):
+        self.owner = owner
+        self.ratn = ratn
+        self.ratd = ratd
+        self.expiry_round = expiry_round
+        self.min_trade = min_trade
+        self.max_fee = max_fee
+
+    def get_program(self):
+        """
+        Return a byte array to be used in LogicSig.
+        """
+        pass
+
+    def get_swap_assets_transctions(self, amount: int, contract: bytes,
+                                    secret_key: bytes):
+        """
+        Return a group transactions array which transfer funds according to
+        the contract's ratio.
+
+        Args:
+        ----------
+        amount : int
+            the amount of assets to be sent
+        contract : bytes
+            the contract containg information, should be recived from payer
+        seceret_key : bytes
+            the secret key to sign the contract
+        """
+
+
+def put_uvarint(buf, x):
+    i = 0
+    while x >= 0x80:
+        buf.append((x & 0xFF) | 0x80)
+        x >>= 7
+        i += 1
+
+    buf.append(x & 0xFF)
+    return i + 1
+
+
+def inject(orig, offsets, values, values_types):
+    # make sure we have enough values
+    assert len(offsets) == len(values) == len(values_types)
+
+    res = orig[:]
+
+    def replace(arr, new_val, offset, place_holder_length):
+        return arr[:offset] + new_val + arr[offset+place_holder_length:]
+
+    for i in range(len(offsets)):
+        val = values[i]
+        val_type = values_types[i]
+        dec_len = 0
+
+        if val_type == int:
+            buf = []
+            dec_len = put_uvarint(buf, val)
+            val = bytes(buf)
+            res = replace(res, val, offsets[i], 1)
+
+        elif val_type == "address":
+            val = encoding.decode_address(val)
+            res = replace(res, val, offsets[i], 32)
+
+        elif val_type == "base64":
+            val = bytes(base64.b64decode(val))
+            buf = []
+            dec_len = put_uvarint(buf, len(val)) + len(val)
```

Comment:
dec_len = put_uvarint(buf, len(val)) + len(val) **-1** ?
Since the base64 placeholder is 2, and 1 is subtracted in line 432.